extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/rules/controllers.rules.js

```diff
@@ -295,6 +296,173 @@ angular.module('PaperUI.controllers.rules', [ 'PaperUI.controllers.extension' ])
+                visibility : 'VISIBLE'
+            });
+            if (mtype != 'trigger' || type == 'trigger') {
+                $scope.moduleData = modules;
+            }
+            if (callback) {
+                $scope.triggerData = modules;
+                if ($scope.module) {
+                    callback();
+                }
+            } else {
+                getModulesByType('trigger', setConfigurations);
+            }
+
+        });
+    }
+    $scope.id = module.id;
+    $scope.type = type;
+    $scope.description = '';
+    $scope.module = '';
+    $scope.step = 1;
+    $scope.editMode = false;
+    $scope.configuration = {};
+    $scope.parameters = [];
+    var originalConfiguration = {};
+    $scope.items = [];
+
+    $scope.selectChip = function(chip, textAreaName) {
+        var textArea = $("textarea[name=" + textAreaName + "]")[0];
+        var textBefore = textArea.value.substring(0, textArea.selectionStart);
+        var textAfter = textArea.value.substring(textArea.selectionStart, textArea.value.length);
+        $scope.configuration[textAreaName] = textBefore + chip.name + textAfter;
+    }
+
+    var setConfigurations = function() {
+        if ($scope.moduleData) {
+            var params = filterByUid($scope.moduleData, $scope.module);
+            if (params && params.length > 0) {
+                $scope.parameters = configService.getRenderingModel(params[0].configDescriptions);
+            }
+            var hasScript = false;
+            angular.forEach($scope.parameters, function(value) {
+
+                sharedProperties.updateParams(value);
+                hasScript = $.grep(value.parameters, function(parameter) {
+                    return parameter.context == 'script';
+                }).length > 0;
+            });
+
+            var index = sharedProperties.searchArray(sharedProperties.getModuleArray(type), $scope.id);
+            if (index != -1) {
+                $scope.configuration = configService.convertValues(sharedProperties.getModuleArray(type)[index].configuration);
+                angular.copy($scope.configuration, originalConfiguration);
+            }
+            $scope.configuration = configService.setConfigDefaults($scope.configuration, $scope.parameters);
+            if (hasScript && type != 'trigger') {
+                var triggers = sharedProperties.getModuleArray('trigger');
+                angular.forEach(triggers, function(trigger, i) {
+                    var moduleType = filterByUid($scope.triggerData, trigger.type);
+                    if (moduleType && moduleType.length > 0) {
+                        $scope.items = $scope.items.concat(moduleType[0].outputs);
+                    }
+                });
+                if (type == 'action') {
+                    var actions = sharedProperties.getModuleArray('action');
+                    for (var i = 0; i < sharedProperties.searchArray(actions, $scope.id); i++) {
```

Comment:
Performance-wise it would be better to move `sharedProperties.searchArray(actions, $scope.id);` to a separate variable:

```js
var actionsArray = sharedProperties.searchArray(actions, $scope.id);
for (var i = 0; i < actionsArray; i++) {
   ...
}
```

This way it won't recalculate it each time.
Also, consider using Array.forEach here.